clang/include/clang/Basic/Builtins.td

```diff
@@ -839,6 +839,12 @@ def BuiltinAssumeAligned : Builtin {
   let Prototype = "void*(void const*, size_t, ...)";
 }
 
+def BuiltinAssumeDereferenceable : Builtin {
+  let Spellings = ["__builtin_assume_dereferenceable"];
+  let Attributes = [NoThrow, Const, Constexpr];
```

Comment:
This says `Constexpr`, but there's no implementation for that that, and of course no tests.

Reply:
I think it would be really nice to make this `constexpr`, since we'd have to jump through hoops otherwise to use this in e.g. `string` (yes, I already have ideas).

Reply:
The constexpr implementation should be pretty easy, since this is effectively a no-op, right? Though perhaps you still have to evaluate the arguments.

Reply:
I think we should make sure that the claimed range is actually dereferenceable, just like we check in `__builtin_assume_aligned` that that pointer is actually guaranteed to be aligned properly. I don't know how complicated that would be.

Reply:
WDYT about adding `constexpr` support separately? One use case I am interested in is using this with some libc++ data types like `std::vector` :)

Reply:
Not sure how you want to use it with `vector`, but I'd be fine with making `constexpr` support a follow-up patch. Given that `vector` is `constexpr` I'd really like to see the support implemented before adding it to libc++ though, since it'd require quite a bit of working around non-constexpr support otherwise.